jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java

```diff
@@ -126,14 +122,40 @@ public ArrayTrie()
      * store "bar" and "bat".
      */
     @SuppressWarnings("unchecked")
-    public ArrayTrie(int capacity)
+    ArrayTrie(int capacity)
     {
         super(true);
+        capacity++;
         _value = (V[])new Object[capacity];
-        _rowIndex = new char[capacity * 32];
+        _rowIndex = new char[capacity * ROW_SIZE];
         _key = new String[capacity];
     }
 
+    @SuppressWarnings("unchecked")
+    ArrayTrie(Map<String, V> initialValues)
+    {
+        super(true);
+        int capacity = requiredCapacity(initialValues.keySet()) + 1;
+        _value = (V[])new Object[capacity];
+        _rowIndex = new char[capacity * ROW_SIZE];
+        _key = new String[capacity];
+        for (Map.Entry<String, V> entry : initialValues.entrySet())
+        {
+            if (!put(entry.getKey(), entry.getValue()))
+                throw new AssertionError("Invalid capacity calculated (" + capacity + ") at '" + entry + "' for " + initialValues);
+        }
+    }
+
+    static int requiredCapacity(Set<String> keys)
+    {
+        int total = 0;
+        for (String key : keys)
+        {
+            total += key.length() + 1;
```

Comment:
This is not the correct algorithm as common roots of keys share nodes.  See the one I posted earlier

Reply:
I know, but the one you posted earlier is partially broken, so I reverted to this simple impl to get the API right and the tests passing.

Reply:
oh do tell what is broken! that sounds like fun!!!!
But at least comment for now that this is an over approximation.

Reply:
IIRC it did not calculate the correct size for `HTTP2Cipher.__blackCiphers`, it underestimated it.

Reply:
Ah  I think I have it this time... and it is simpler:
```java

    public static int requiredCapacity(Set<String> keys)
    {
        List<String> list = new ArrayList<>(keys);
        Collections.sort(list);
        return requiredCapacity(list, 0, list.size(), 0);
    }

    private static int requiredCapacity(List<String> keys, int offset, int length, int index)
    {
        if (length == 0)
            return 0;

        int required = 0;

        Character c = null;
        for (int i = 0; i < length; i++)
        {
            String k = keys.get(offset + i);
            if (k.length() <= index)
                continue;
            char n = k.charAt(index);
            if (c == null)
            {
                required++;
                c = n;
                offset += i;
                length -= i;
            }
            else if (c != n)
            {
                required +=  requiredCapacity(keys, offset, i, index + 1) + 1;
                offset += i;
                length -= i;
                c = n;
                i = 1;
            }
        }

        if (c != null)
            required += requiredCapacity(keys, offset, length, index + 1);
        return required;
    }

    @Test
    public void testRequiredCapacity()
    {
        assertThat(requiredCapacity(Set.of("A")), is(1));
        assertThat(requiredCapacity(Set.of("AB")), is(2));
        assertThat(requiredCapacity(Set.of("ABC")), is(3));
        assertThat(requiredCapacity(Set.of("ABC", "XYZ")), is(6));
        assertThat(requiredCapacity(Set.of("A00", "A11")), is(5));
        assertThat(requiredCapacity(Set.of("A00", "A01", "A10", "A11")), is(7));
        assertThat(requiredCapacity(Set.of("A", "AB")), is(2));
        assertThat(requiredCapacity(Set.of("A", "ABC")), is(3));
        assertThat(requiredCapacity(Set.of("A", "ABCD")), is(4));
        assertThat(requiredCapacity(Set.of("AB", "ABC")), is(3));
        assertThat(requiredCapacity(Set.of("ABC", "ABCD")), is(4));
        assertThat(requiredCapacity(Set.of("ABC", "ABCDEF")), is(6));
        assertThat(requiredCapacity(Set.of("AB", "A")), is(2));
        assertThat(requiredCapacity(Set.of("ABC", "ABCDEF")), is(6));
        assertThat(requiredCapacity(Set.of("ABCDEF", "ABC")), is(6));
        assertThat(requiredCapacity(Set.of("ABC", "ABCDEF", "ABX")), is(7));
        assertThat(requiredCapacity(Set.of("ABCDEF", "ABC", "ABX")), is(7));
        assertThat(requiredCapacity(Set.of("ADEF", "AQPR4", "AQZ")), is(9));
        assertThat(requiredCapacity(Set.of("111", "ADEF", "AQPR4", "AQZ", "999")), is(15));
    }
```

Reply:
This one still fails:

```java
  assertThat(ArrayTrie.requiredCapacity(Set.of("utf-8", "utf8", "utf-16", "utf16", "iso-8859-1", "iso_8859_1")), is(27));
```

Reply:
So much fun!!!!:
```java

    
    public static int requiredCapacity(Set<String> keys)
    {
        List<String> list = new ArrayList<>(keys);
        Collections.sort(list);
        return requiredCapacity(list, 0, list.size(), 0);
    }

    private static int requiredCapacity(List<String> keys, int offset, int length, int index)
    {
        if (length == 0)
            return 0;

        int required = 0;

        Character c = null;
        for (int i = 0; i < length; i++)
        {
            String k = keys.get(offset + i);
            if (k.length() <= index)
                continue;
            char n = k.charAt(index);
            if (c == null)
                required++;
            else if (c != n)
                required +=  requiredCapacity(keys, offset, i, index + 1) + 1;
            else
                continue;
            c = n;
            offset += i;
            length -= i;
            i = 0;
        }

        if (c != null)
        {
            required += requiredCapacity(keys, offset, length, index + 1);
        }
        return required;
    }

    @Test
    public void testRequiredCapacity()
    {
        assertThat(requiredCapacity(Set.of("ABC")), is(3));
        assertThat(requiredCapacity(Set.of("ABC", "XYZ")), is(6));
        assertThat(requiredCapacity(Set.of("A00", "A11")), is(5));
        assertThat(requiredCapacity(Set.of("A00", "A01", "A10", "A11")), is(7));
        assertThat(requiredCapacity(Set.of("A", "AB")), is(2));
        assertThat(requiredCapacity(Set.of("A", "ABC")), is(3));
        assertThat(requiredCapacity(Set.of("A", "ABCD")), is(4));
        assertThat(requiredCapacity(Set.of("AB", "ABC")), is(3));
        assertThat(requiredCapacity(Set.of("ABC", "ABCD")), is(4));
        assertThat(requiredCapacity(Set.of("ABC", "ABCDEF")), is(6));
        assertThat(requiredCapacity(Set.of("AB", "A")), is(2));
        assertThat(requiredCapacity(Set.of("ABC", "ABCDEF")), is(6));
        assertThat(requiredCapacity(Set.of("ABCDEF", "ABC")), is(6));
        assertThat(requiredCapacity(Set.of("ABC", "ABCDEF", "ABX")), is(7));
        assertThat(requiredCapacity(Set.of("ABCDEF", "ABC", "ABX")), is(7));
        assertThat(requiredCapacity(Set.of("ADEF", "AQPR4", "AQZ")), is(9));
        assertThat(requiredCapacity(Set.of("111", "ADEF", "AQPR4", "AQZ", "999")), is(15));
        assertThat(requiredCapacity(Set.of("utf-16", "utf-8")), is(7));
        assertThat(requiredCapacity(Set.of("utf-16", "utf-8", "utf16", "utf8")), is(10));
        assertThat(requiredCapacity(Set.of("utf-8", "utf8", "utf-16", "utf16", "iso-8859-1", "iso_8859_1")), is(27));
    }
```

Reply:
final version:
```java

    /**
     * Calculate required Trie capacity in nodes of a tree decomposition of the keys.
     * For example given the keys:
     * <ul>
     *     <li>utf_16</li>
     *     <li>utf_8</li>
     *     <li>utf16</li>
     *     <li>utf8</li>
     * </ul>
     * The tree has 10 nodes as follows:
     * <pre>
     *                     1 - 6
     *                   /
     *                 _ - 8
     *               /
     *     u - t - f - 1 - 6
     *               \
     *                 8
     * </pre>
     * @param keys The keys to be put in a Trie
     * @return The capacity in nodes of a tree decomposition
     */
    public static int requiredCapacity(Set<String> keys)
    {
        List<String> list = new ArrayList<>(keys);
        Collections.sort(list);
        return requiredCapacity(list, 0, list.size(), 0);
    }

    /**
     * Calculate required Trie capacity in nodes of a sub-tree decomposition of the keys.
     * @param keys The keys to calculate the capacity for
     * @param offset The offset of the first key to be considered
     * @param length The number of keys to be considered
     * @param index The character to be considered
     * @return The capacity in tree nodes of the substree
     */
    private static int requiredCapacity(List<String> keys, int offset, int length, int index)
    {
        int required = 0;

        // Examine all the keys in the subtree
        Character nodeChar = null;
        for (int i = 0; i < length; i++)
        {
            String k = keys.get(offset + i);

            // If the key is shorter than our current index then ignore it
            if (k.length() <= index)
                continue;

            // Get the character at the index of the current key
            char c = k.charAt(index);

            // If the character is the same as the current node, then we are
            // still in the current node and need to continue searching for the
            // next node or the end of the keys
            if (nodeChar != null && c == nodeChar)
                continue;

            // The character is a new node, so increase required by 1
            required++;

            // if we had a previous node, then add the required nodes for the subtree under it.
            if (nodeChar != null)
                required +=  requiredCapacity(keys, offset, i, index + 1);

            // set the char for the new node
            nodeChar = c;

            // reset the offset, length and index to continue iteration from the start of the new node
            offset += i;
            length -= i;
            i = 0;
        }

        // If we finish the iteration with a nodeChar, then we must add the required nodes for the subtree under it.
        if (nodeChar != null)
            required += requiredCapacity(keys, offset, length, index + 1);

        return required;
    }

    @Test
    public void testRequiredCapacity()
    {
        assertThat(requiredCapacity(Set.of("")), is(0));
        assertThat(requiredCapacity(Set.of("ABC", "")), is(3));
        assertThat(requiredCapacity(Set.of("ABC")), is(3));
        assertThat(requiredCapacity(Set.of("ABC", "XYZ")), is(6));
        assertThat(requiredCapacity(Set.of("A00", "A11")), is(5));
        assertThat(requiredCapacity(Set.of("A00", "A01", "A10", "A11")), is(7));
        assertThat(requiredCapacity(Set.of("A", "AB")), is(2));
        assertThat(requiredCapacity(Set.of("A", "ABC")), is(3));
        assertThat(requiredCapacity(Set.of("A", "ABCD")), is(4));
        assertThat(requiredCapacity(Set.of("AB", "ABC")), is(3));
        assertThat(requiredCapacity(Set.of("ABC", "ABCD")), is(4));
        assertThat(requiredCapacity(Set.of("ABC", "ABCDEF")), is(6));
        assertThat(requiredCapacity(Set.of("AB", "A")), is(2));
        assertThat(requiredCapacity(Set.of("ABC", "ABCDEF")), is(6));
        assertThat(requiredCapacity(Set.of("ABCDEF", "ABC")), is(6));
        assertThat(requiredCapacity(Set.of("ABC", "ABCDEF", "ABX")), is(7));
        assertThat(requiredCapacity(Set.of("ABCDEF", "ABC", "ABX")), is(7));
        assertThat(requiredCapacity(Set.of("ADEF", "AQPR4", "AQZ")), is(9));
        assertThat(requiredCapacity(Set.of("111", "ADEF", "AQPR4", "AQZ", "999")), is(15));
        assertThat(requiredCapacity(Set.of("utf-16", "utf-8")), is(7));
        assertThat(requiredCapacity(Set.of("utf-16", "utf-8", "utf16", "utf8")), is(10));
        assertThat(requiredCapacity(Set.of("utf-8", "utf8", "utf-16", "utf16", "iso-8859-1", "iso_8859_1")), is(27));
    }
```

Reply:
Works perfectly fine! Integrated.